baseapp/abci.go

```diff
@@ -118,140 +115,323 @@ func (app *BaseApp) InitChain(req abci.RequestInitChain) (res abci.ResponseInitC
+	// not.
+	//
+	// Ref: https://github.com/cosmos/cosmos-sdk/issues/12272
+	err = app.snapshotManager.Restore(snapshot)
+	switch {
+	case err == nil:
+		return &abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_ACCEPT}, nil
+
+	case errors.Is(err, snapshottypes.ErrUnknownFormat):
+		return &abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT_FORMAT}, err
+
+	case errors.Is(err, snapshottypes.ErrInvalidMetadata):
+		app.logger.Error(
+			"rejecting invalid snapshot",
+			"height", req.Snapshot.Height,
+			"format", req.Snapshot.Format,
+			"err", err,
+		)
+		return &abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT}, err
+
+	default:
+		app.logger.Error(
+			"failed to restore snapshot",
+			"height", req.Snapshot.Height,
+			"format", req.Snapshot.Format,
+			"err", err,
+		)
+
+		// We currently don't support resetting the IAVL stores and retrying a
+		// different snapshot, so we ask CometBFT to abort all snapshot restoration.
+		return &abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_ABORT}, nil
+	}
+}
+
+// ApplySnapshotChunk implements the ABCI interface. It delegates to app.snapshotManager if set.
+func (app *BaseApp) ApplySnapshotChunk(_ context.Context, req *abci.RequestApplySnapshotChunk) (*abci.ResponseApplySnapshotChunk, error) {
+	if app.snapshotManager == nil {
+		app.logger.Error("snapshot manager not configured")
+		return &abci.ResponseApplySnapshotChunk{Result: abci.ResponseApplySnapshotChunk_ABORT}, nil
 	}
 
-	gasMeter := app.getBlockGasMeter(app.deliverState.ctx)
+	// TODO: Examine non-ACCEPT return paths and if errors should be returned or
+	// not.
+	//
+	// Ref: https://github.com/cosmos/cosmos-sdk/issues/12272
+	_, err := app.snapshotManager.RestoreChunk(req.Chunk)
+	switch {
+	case err == nil:
+		return &abci.ResponseApplySnapshotChunk{Result: abci.ResponseApplySnapshotChunk_ACCEPT}, nil
 
-	app.deliverState.ctx = app.deliverState.ctx.
-		WithBlockGasMeter(gasMeter).
-		WithHeaderHash(req.Hash).
-		WithConsensusParams(app.GetConsensusParams(app.deliverState.ctx)).
-		WithVoteInfos(req.LastCommitInfo.GetVotes())
+	case errors.Is(err, snapshottypes.ErrChunkHashMismatch):
+		app.logger.Error(
+			"chunk checksum mismatch; rejecting sender and requesting refetch",
+			"chunk", req.Index,
+			"sender", req.Sender,
+			"err", err,
+		)
+		return &abci.ResponseApplySnapshotChunk{
+			Result:        abci.ResponseApplySnapshotChunk_RETRY,
+			RefetchChunks: []uint32{req.Index},
+			RejectSenders: []string{req.Sender},
+		}, err
 
-	if app.checkState != nil {
-		app.checkState.ctx = app.checkState.ctx.
-			WithBlockGasMeter(gasMeter).
-			WithHeaderHash(req.Hash)
+	default:
+		app.logger.Error("failed to restore snapshot", "err", err)
+		return &abci.ResponseApplySnapshotChunk{Result: abci.ResponseApplySnapshotChunk_ABORT}, nil
 	}
+}
+
+func (app *BaseApp) legacyBeginBlock(req *abci.RequestFinalizeBlock) sdk.LegacyResponseBeginBlock {
```

Comment:
i dont think we need this anymore, streamer should work off of finalise block

Reply:
Agree, the only module that has a BeginBlocker that uses the legacy `abci.RequestBeginBlock` is x/evidence. To stop using it we need to pass ByzantineValidators/Misbehaviors through the context (or something else, but maybe the context is fine for now).

Reply:
We would also need to modify the module manager beginblock to return a list of `abci.Event` instead of `abci.ResponseBeginBlock`. I can add that to https://github.com/cosmos/cosmos-sdk/pull/15744

Reply:
Just reached the end of the PR, disregard my comments above, you already thought of this 😅

Reply:
@tac0turtle I'm assuming you're referring to the arguments/function signature and not the need for `legacyBeginBlock` more broadly, which is needed.

Yes, I agree, for both `BeginBlock` and `EndBlock` we can simply use `Context` as inputs. However, they both have returns values that are needed for CometBFT.

* `BeginBlock` will need to return events
* `EndBlock` will need to return events, validator updates, and consensus params (ref: https://github.com/cometbft/cometbft/blob/v0.34.27/proto/tendermint/abci/types.proto#L236-L241)

Reply:
Feel free to push changes to this PR @facundomedica or @tac0turtle 👍...I'm moving onto vote extensions